### PR TITLE
compute_ctl: Break up main()

### PR DESCRIPTION
**Pull request by sharnoff**
## Problem

compute_ctl's `main()` is ["pretty gnarly"](https://github.com/neondatabase/neon/pull/7434#discussion_r1574828357). This makes changes to control flow quite difficult.

I hit this while working on #7434 and had an idea of how it could be improved, thus this PR.

## Summary of changes

First commit: Small changes to reduce the diff of the second commit.

Second commit: Breaks up `main()` into a handful of component functions, aligned with distinct sections of the existing control flow. Each new function gets the previous one's return values as its arguments. The new functions destructure their args so the names are all the same.

Third commit: Move the tracing startup context guard out of where it was previously used, to make things more simple. Taken from #7600.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: compute_tools/src/bin/compute_ctl.rs [hunk 1]
```diff
@@ -256,6 +317,19 @@ fn main() -> Result<()> {
...
+    Ok(WaitSpecResult { compute, http_port })
+}
+
+struct WaitSpecResult {
+    compute: Arc<ComputeNode>,
+    // passed through from ProcessCliResult
+    http_port: u16,
+}
+
+fn start_postgres(
```
Comment (tristan957) [inline compute_tools/src/bin/compute_ctl.rs:329]: This function contains quite a bit of things that are unrelated to starting Postgres. Could we chop it up as well?

File: compute_tools/src/bin/compute_ctl.rs [hunk 2]
```diff
@@ -349,11 +423,47 @@ fn main() -> Result<()> {
...
+    pg: Option<PostgresHandle>,
+    startup_context_guard: Option<opentelemetry::ContextGuard>,
+) -> Result<WaitPostgresResult> {
     // Wait for the child Postgres process forever. In this state Ctrl+C will
     // propagate to Postgres and it will be shut down as well.
     let mut exit_code = None;
     if let Some((mut pg, logs_handle)) = pg {
         // Startup is finished, exit the startup tracing span
+        // TODO: Probably easier to drop startup_context_guard outside this function. It's here
+        // right now because keeping it here reduced the size of the diff.
```
Comment (tristan957) [inline compute_tools/src/bin/compute_ctl.rs:466]: Doing this in a future commit within this PR?

File: compute_tools/src/bin/compute_ctl.rs [hunk 3]
```diff
@@ -370,6 +480,26 @@ fn main() -> Result<()> {
...
     }
 
+    Ok(WaitPostgresResult { exit_code })
+}
+
+struct WaitPostgresResult {
+    exit_code: Option<i32>,
+}
+
+fn cleanup_and_exit(
```
Comment (tristan957) [inline compute_tools/src/bin/compute_ctl.rs:490]: Splitting this function up into smaller logical chunks would also be great.

File: compute_tools/src/bin/compute_ctl.rs [hunk 4]
```diff
@@ -202,6 +222,47 @@ fn main() -> Result<()> {
...
+        spec,
+        live_config_allowed,
+    };
+
+    // TODO: Move startup_context_guard out of this function. It's here right now only because
+    // that's where it was before, and moving it would've made the diff big.
+    Ok((startup_context_guard, result))
+}
+
+struct ProcessCliResult<'clap> {
```
Comment (tristan957) [inline compute_tools/src/bin/compute_ctl.rs:242]: Doesn't clap have a way to to parse a command line into a struct? Seems like that would be the best solution to cleaning up the command line code, imo.

File: compute_tools/src/bin/compute_ctl.rs (hunk 2 above)
Comment (sharnoff) [reply compute_tools/src/bin/compute_ctl.rs:466]: Not sure what semantics are intended around the startup context — initially decided to leave it for now pending any discussion that should be had on that

But otherwise happy to do so.

File: compute_tools/src/bin/compute_ctl.rs (hunk 4 above)
Comment (sharnoff) [reply compute_tools/src/bin/compute_ctl.rs:242]: Yup- I quite like its derive API. I figured that's a non-trivial change, only want to tackle that in this PR if there's +1s from compute team.

File: compute_tools/src/bin/compute_ctl.rs (hunk 3 above)
Comment (sharnoff) [reply compute_tools/src/bin/compute_ctl.rs:490]: Makes sense, yeah, will split it up a little bit more — hesitant to go much further though because IMO logically grouping actions by the "phase" of compute_ctl execution provides a lot of context when reading the code, and I worry that smaller chunks may leave `main()` as the only place with adequate context.

File: compute_tools/src/bin/compute_ctl.rs (hunk 1 above)
Comment (sharnoff) [reply compute_tools/src/bin/compute_ctl.rs:329]: Same hesitancy here as with `cleanup_and_exit`. I'll see what I can do.

I think the "right" way to do this is probably to have some abstraction around starting/stopping vm-monitor stuff, so that the usage in `compute_ctl.rs` is just one-liners, but I considered that outside the scope of this PR.

Happy to include it as a separate commit if desired.

File: compute_tools/src/bin/compute_ctl.rs (hunk 4 above)
Comment (tristan957) [reply compute_tools/src/bin/compute_ctl.rs:242]: I am definitely a +1 on it :), but you also don't need to tackle it as part of this PR. I could get to it.

File: compute_tools/src/bin/compute_ctl.rs (hunk 2 above)
Comment (hlinnaka) [reply compute_tools/src/bin/compute_ctl.rs:466]: This function doesn't do anything with it, except drop it, so +1 for moving it to the caller.

Comment (hlinnaka) [reply compute_tools/src/bin/compute_ctl.rs:466]: See https://github.com/neondatabase/neon/pull/7600 for suggestion on how that would look

Comment (sharnoff) [reply compute_tools/src/bin/compute_ctl.rs]: Implemented #7600 via 8284db01f6ff12415421f5f820611143b66bc9f9 and ffe65ef84f386d3bc4b34cf41e2c3212aee624c9. Some adjustments made to avoid moving things around (because it makes the diff big, which makes merge conflicts harder to resolve)

File: compute_tools/src/bin/compute_ctl.rs (hunk 3 above)
Comment (sharnoff) [reply compute_tools/src/bin/compute_ctl.rs]: Split it up further in 483559ae4741af296b68f1850d800b8e2eb37e6d

File: compute_tools/src/bin/compute_ctl.rs (hunk 1 above)
Comment (sharnoff) [reply compute_tools/src/bin/compute_ctl.rs:358]: Some of this will be handled by #7601 - other than that, I think it's just separating out the vm-monitor stuff, which I'm hesitant to do. Maybe better to name it something other than `start_postgres`... `start_compute_tasks` ? not sure. WDYT?

**Review by hlinnaka — APPROVED** [review]

**Comment by sharnoff** [comment]
Planning to merge in 24h or so, if no objections (cc @tristan957).

**Review by tristan957 — APPROVED** [review]
